dist/bearhug-angular.js

```diff
@@ -0,0 +1,207 @@
+            var opts = angular.copy(options);
+            if (!angular.isFunction(opts.authenticationFunction)) {
+                throw new Error("No authentication function specified. How can bearhug authenticate?");
+            } else {
+                return {
+                    options: opts,
+                    authenticate: authenticate,
+                    getToken: bearhugStorage.getToken,
+                    getBearer: bearhugStorage.getBearer
+                };
+            }
+            function authenticate(requestConfig) {
+                return bearhugAuthenticator.authenticate(opts, requestConfig);
+            }
+        }];
+    }
+    angular.module("talis.bearhug").factory("bearhugAuthenticator", bearhugAuthenticator);
+    function bearhugAuthenticator($http, $injector, $q, bearhugStorage) {
+        var authenticationDeferred;
+        return {
+            authenticate: authenticate
+        };
+        function authenticate(options) {
+            var opts = angular.isObject(options) ? options : {};
+            if (!angular.isFunction(opts.authenticationFunction)) {
+                throw new Error("No authentication function available to bearhugAuthenticator");
+            } else if (authenticationDeferred) {
+                return authenticationDeferred.promise;
+            } else {
+                authenticationDeferred = $q.defer();
+                var authPromise = opts.authenticationFunction($injector).then(function(tokenFromResponse) {
+                    bearhugStorage.setToken(tokenFromResponse);
+                    var token = bearhugStorage.getToken();
+                    if (token) {
+                        authenticationDeferred.resolve(token);
+                    } else {
+                        authenticationDeferred.reject(new Error("Token response is not a string: " + angular.toJson(tokenFromResponse)));
+                    }
+                }).catch(authenticationDeferred.reject);
+                authPromise.finally(function() {
+                    authenticationDeferred = void 0;
+                });
+                return authenticationDeferred.promise;
+            }
+        }
+    }
+    angular.module("talis.bearhug").factory("bearhugInterceptor", bearhugInterceptor);
+    function bearhugInterceptor($q, $injector, bearhugStorage) {
+        var authenticationDeferred;
+        return {
+            request: request,
+            response: response,
+            responseError: responseError
+        };
+        function request(requestConfig) {
+            var bearer = bearhugStorage.getBearer();
+            if (bearer && requestConfig && requestConfig.headers) {
+                requestConfig.headers.Authorization = bearer;
+            }
+            return requestConfig;
+        }
+        function response(responseObj) {
+            var bearer = responseObj && responseObj.headers("Authorization");
```

Comment:
It's a [generally accepted style](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md), using function-hoisting to keep public content at the top of files, whilst hiding implementation details lower down.

I'm not sure how I feel about it, given that methods work top-to-bottom regardless.  I'll happily reformat to a linting standard.  I was going to move to ES2015 and eslint in short order, so that it's settled for the project.